docket/tasks.py

```diff
@@ -1,128 +1,283 @@
+
+    threads = []
+    # Query each instance concurrently
+    for instance in _INSTANCES:
+        from threading import Thread
+        thread = Thread(target=_requester, args=(query, instance, headers, query.job_path))
+        thread.start()
+        threads.append(thread)
+
+    # Wait until all threads complete.
+    alive = True
+    while alive:
+        alive = False
+        for thread in threads:
+            thread.join(1.0)
+            alive = alive or thread.is_alive()
+
+    errors = query.errors
+    status = None
+    if errors:
+        query.error('stenographer', "stenographer queries complete", Query.FAIL)
+        status = "{}: ".format(Query.FAIL) + ', '.join([ "{}:{}".format(*i) for i in errors])
+    else:
+        query.progress('stenographer', "stenographer queries complete", Query.SUCCESS)
+        merge.apply_async(queue='io', kwargs={'query_tuple':query.tupify()})
+        status = Query.SUCCESS
+    query.save()
+    return status
+
+def _requester(query, instance, headers, path):
+    """ Runs in the 'query' worker: performs a request against the instance
+        and writes the response data to disk
+    """
+
+    query.progress(instance['sensor'], "Awaiting idle", Query.RECEIVING)
+    if not _ensure_idle(instance):
+        query.error(instance['sensor'], "idle timeout({})".format(QUERY_TIMEOUT), Query.FAIL)
+        return
+
+    url = "https://%s:%i/query" % (instance['host'], instance['port'])
+    try:
+        query.progress(instance['sensor'], "requesting {}".format(query.query), Query.RECEIVING)
+        rq = requests.post(url, data=query.query, headers=headers,
+                           cert=(instance['cert'], instance['key']),
+                           verify=instance['ca'],
+                           timeout=QUERY_TIMEOUT
+                          )
+        if rq.status_code == requests.codes.bad:
+            error = rq.reason
+            query.error(instance['sensor'], "{} {}".format(rq.status_code, error))
+        elif rq.status_code == requests.codes.ok:
+            query.progress(instance['sensor'], '{} bytes received'.format(len(rq.content)), Query.RECEIVED)
+            out_file = join(path, "{}.pcap".format(instance['sensor']) )
+            with open(out_file, 'wb') as f:
+                f.write(rq.content)
+
+    except requests.exceptions.ConnectTimeout as ex:
+        query.error(instance['sensor'], "Connection Timeout({}) - {}".format(QUERY_TIMEOUT, ex) )
+    except requests.exceptions.ReadTimeout as ex:
+        query.error(instance['sensor'], "Data Timeout({}) - {}".format(QUERY_TIMEOUT, ex))
+    except requests.exceptions.ConnectionError as ex:
+        query.error(instance['sensor'], "Connection Failed - check host:port {}".format(ex))
+        raise ex
+    except requests.exceptions.SSLError as ex:
+        query.error(instance['sensor'], "SSL Failed - check certificate config: {}".format(ex))
+        raise ex
+    return
+
+def _ensure_idle(instance):
```

Comment:
I'm not quite clear on the purpose of querying stats every `idle` time period.

Reply:
We don't want to hammer a stenographer instance that's already busy, even if we aren't the sole user.